### PR TITLE
Write the documentation

### PR DESCRIPTION
**Pull request by Hywan**
- [x] Abstract
- [x] Introduction
- [x] Global workflow
    - [x] Grammar
    - [x] Context
    - [x] Asserter
        - [x] Add functions
- [x] Language transformation
    - [x] Interpreter: language to object model
    - [x] Compiler: object model to PHP
    - [x] Disassembler: object model to language
- [x] Performances
    - [x] Serialize the object model
    - [x] Save and execute PHP code
- [x] Conclusion

### CODE REVIEW

**Comment by Hywan** [comment]
Please @hoaproject/hoackers, time to review the French version!

**Comment by Hywan** [comment]
You can use https://github.com/hoaproject/Hoathis-Documentation to generate the documentation locally.

**Comment by Hywan** [comment]
Use the following `composer.json`:

```json
{
    "require": {
        "hoa/ruler": "dev-documentation",
        "hoathis/documentation": "dev-master"
    },
    "repositories": [
        {
            "type": "vcs",
            "url": "https://github.com/Hywan/Ruler"
        }
    ]
}
```

Then:

```sh
$ composer install
$ vendor/bin/hoa hoathis documentation:generate
```

Finally, open the printed link.

(`hoathis/documentation` is under release, so use `dev-master` for now).

File: Documentation/Fr/Index.xyl [hunk 1]
```diff
@@ -0,0 +1,729 @@
...
+  langage utilisé pour exprimer une règle n'est pas le langage utilisé pour
+  développer le programme. Un exemple encore plus évident avec l'utilisation
+  d'une règle pour filtrer des éléments : un élément est accepté si « <q>son
+  groupe est <em lang="en">customer</em> ou <em lang="en">guest</em> et que son
+  nombre de points est supérieur à 30</q> ». Cette règle peut très bien être
+  écrite par un utilisateur via une interface en ligne de commande pour filtrer
+  des résultats d'une base de données ou de logs.</p>
+  <p>Il est important de comprendre que les règles doivent être écrites dans un
+  <strong>langage dédié</strong>. Néanmoins, l'usage qui est fait des règles est
+  très <strong>vaste</strong> et <strong>imprédictible</strong>. C'est pourquoi
```
Comment (vonglasow) [inline Documentation/Fr/Index.xyl]: "C'est pourquoi, il est primordiale d'avoir des règles souples et extensibles dans leur syntaxe."

File: Documentation/Fr/Index.xyl [hunk 2]
```diff
@@ -0,0 +1,729 @@
...
+  il est primordiale que les règles soient souples et
+  <strong>extensibles</strong> dans leur syntaxe. Par exemple, il doit être
+  permis d'ajouter des opérateurs ou des fonctions : « <q>tous les clients de
+  l'hôtel avec un pass Gold auront une réduction de 10%</q> ». Le « pass Gold »
+  peut être un opérateur ou une fonction spécifique au métier du programme
+  concerné.</p>
+  <p>Le langage utilisé par la bibliothèque <code>Hoa\Ruler</code> pour décrire
+  des règles respecte ces contraintes d'extensibilité. Les règles ne seront pas
+  proches du langage humain mais resterons <strong>naturelles</strong> à lire.
+  Si nous reprenons l'exemple de la règle « <q>son groupe est
```
Comment (vonglasow) [inline Documentation/Fr/Index.xyl]: s/et que son nombre de points/et son nombre de points/

File: Documentation/Fr/Index.xyl [hunk 3]
```diff
@@ -0,0 +1,729 @@
...
+  <em lang="en">customer</em> ou <em lang="en">guest</em> et que son nombre de
+  points est supérieur à 30</q> », elle s'écrira : <code><em>group</em> in
+  ["customer", "guest"] and <em>points</em> > 30</code>.  Les éléments
+  <code><em>group</em></code> et <code><em>points</em></code> sont des variables
+  de la règle. Leurs valeurs seront définies par un contexte.</p>
+  <p>D'une point de vue plus formel, une règle est un prédicat, c'est à dire que
```
Comment (vonglasow) [inline Documentation/Fr/Index.xyl]: D'un point de vue

File: Documentation/Fr/Index.xyl [hunk 4]
```diff
@@ -0,0 +1,729 @@
...
+  son résultat est toujours un booléen : <code class="language-php">true</code>
+  ou <code class="language-php">false</code>. Comme ces règles sont destinées à
+  être manipulées (modifiées) et exécutées, la bibliothèque
+  <code>Hoa\Ruler</code> propose plusieurs outils pour travailler efficacement
```
Comment (vonglasow) [inline Documentation/Fr/Index.xyl]: propose plusieurs outils,

File: Documentation/Fr/Index.xyl [hunk 5]
```diff
@@ -0,0 +1,729 @@
...
+
+      paper.link.between(rule, vinterpreter, '');
+      paper.link.between(vinterpreter, model, '');
+      paper.link.between(model, vasserter, '');
+      paper.link.between(model, vcompiler, '');
+      paper.link.between(model, vdisassembly, '');
+  });
+  </script>
+
+  <p>Tout d'abord, la règle est interprété par une interpréteur pour être
```
Comment (vonglasow) [inline Documentation/Fr/Index.xyl]: par un interpréteur

**Comment by vonglasow** [comment]
Except the few remarques it look's ok for me ;) :+1:

File: Documentation/Fr/Index.xyl [hunk 6]
```diff
@@ -0,0 +1,729 @@
+<?xml version="1.0" encoding="utf-8"?>
+
+<overlay xmlns="http://hoa-project.net/xyl/xylophone">
+<yield id="chapter">
+
+  <p>Les <strong>règles métiers</strong> (comme « <q>tous les clients qui ont
+  dépensé plus de 100€ en une fois reçoivent une réduction de 10% sur le
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: leur prochain achat

File: Documentation/Fr/Index.xyl [hunk 7]
```diff
@@ -0,0 +1,729 @@
...
+
+  <h2 id="Table_of_contents">Table des matières</h2>
+
+  <tableofcontents id="main-toc" />
+
+  <h2 id="Introduction" for="main-toc">Introduction</h2>
+
+  <p>La logique métier est très différente de la logique Informatique. « <q>Tous
+  les clients qui ont dépensé plus de 100€ en une fois reçoivent une réduction
+  de 10% sur le prochain achat</q> ». Cette règle permet d'accéder à certaines
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: leur prochain achat

File: Documentation/Fr/Index.xyl [hunk 8]
```diff
@@ -0,0 +1,729 @@
...
+  des variables dont les valeurs sont définies par le <strong>contexte</strong>.
+  Le contexte peut contenir des valeurs scalaires, des tableaux ou même des
+  fonctions et des objets. Enfin, l'<strong>asserteur</strong> associe le
+  contexte à la règle pour pouvoir l'exécuter et obtenir un résultat. Nous
+  rappelons que ce résultat est nécessairement un booléen. C'est alors un
+  prédicat.</p>
+  <p>Le contexte est représenté par la classe <code>Hoa\Ruler\Context</code>.
+  L'asserteur est représenté par la classe
+  <code>Hoa\Ruler\Visitor\Asserter</code>. Nous pouvons utiliser la méthode
+  <code>Hoa\Ruler\Ruler::assert</code> qui facilite son utilisation. Ainsi :</p>
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: utiliser ... qui facilite son utilisation // 2x utiliser ;-)
Nous pouvons employer ... qui facilite son utilisation

File: Documentation/Fr/Index.xyl [hunk 9]
```diff
@@ -0,0 +1,729 @@
...
+var_dump(
+    $ruler->assert($rule, $context)
+);
+
+/**
+ * Will output:
+ *     bool(true)
+ */</code></pre>
+  <p>La règle est définie dans la variable
+  <code class="language-php">$rule</code>. Le contexte, quant à lui, dans la
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: il ne manque pas un verbe dans la phrase "Le contexte, quant à lui, ..."

File: Documentation/Fr/Index.xyl [hunk 10]
```diff
@@ -0,0 +1,729 @@
...
+    </tbody>
+  </table>
+
+  <p>Bien sûr, ces exemples représentent des parties atomiques de la grammaire
+  que nous pouvons combiner. Ainsi :
+  <code><em>userA</em>.allows(<em>groups</em>[<em>groupId</em>][<em>userB</em>])</code>
+  est valide. De même que <code>f(<em>user</em>, <em>points</em> > 7 and
+  <em>points</em> &amp;lt; 42)</code> est également valide.</p>
+  <p>En réalité, les fonctions, les opérateurs de comparaisons et les opérateurs
+  d'appartenances ne sont pas définis par la grammaire mais par l'asserteur
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: appartenance

File: Documentation/Fr/Index.xyl [hunk 11]
```diff
@@ -0,0 +1,729 @@
...
+      var vdisassembly = grid.push(paper.rect(0, 0, 110, 80, 3, 'règle'), 4, 2);
+
+      paper.link.between(rule, vinterpreter, '');
+      paper.link.between(vinterpreter, model, '');
+      paper.link.between(model, vasserter, '');
+      paper.link.between(model, vcompiler, '');
+      paper.link.between(model, vdisassembly, '');
+  });
+  </script>
+  <p>La règle générée peut différée un
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: différer

File: Documentation/Fr/Index.xyl [hunk 12]
```diff
@@ -0,0 +1,729 @@
...
+  objet</h3>
+
+  <p>Une fois le modèle objet obtenu, nous pouvons le sérialiser à l'aide de
+  <a href="http://php.net/serialize">la fonction PHP <code>serialize</code></a>.
+  Nous allons obtenir une chaîne de caractères représentant les instances des
+  objets constituant le modèle objet. Pour obtenir à nouveau le modèle objet et
+  être capable d'y appliquer un asserteur, nous utiliserons
+  <a href="http://php.net/unserialize">la fonction PHP
+  <code>unserialize</code></a>. Le résultat de la sérialisation peut être
+  stocker en base de données à la place des règles. Cela nécessite un peu plus
```
Comment (osaris) [inline Documentation/Fr/Index.xyl]: stocké

**Comment by osaris** [comment]
Very good doc, my comments are in the diff ;-)

File: Documentation/Fr/Index.xyl (hunk 6 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 7 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 1 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 2 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed, two times (the typo appears few lines above).

File: Documentation/Fr/Index.xyl (hunk 3 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 4 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: won't fix, we don't need a comma here, no?

File: Documentation/Fr/Index.xyl (hunk 8 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed, thanks!

File: Documentation/Fr/Index.xyl (hunk 9 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 10 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed, also fixed in the table.

File: Documentation/Fr/Index.xyl (hunk 5 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed.
also s/interprété/interprétée/

Comment (Hywan) [inline Documentation/Fr/Index.xyl]: s/transformé/transformée/

File: Documentation/Fr/Index.xyl (hunk 11 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

File: Documentation/Fr/Index.xyl (hunk 12 above)
Comment (Hywan) [inline Documentation/Fr/Index.xyl]: fixed

**Comment by Hywan** [comment]
Thanks for the precious review!

**Comment by Hywan** [comment]
Now: The translation in English https://github.com/hoaproject/Ruler/pull/56.

**Pull request merged**